### PR TITLE
Abstract transports

### PR DESCRIPTION
**Pull request by runk**
.. was looking into adding another transport (raw stream), however the way different protocols were managed was quite hectic. Doing abstraction as a first step toward that.

### CODE REVIEW

File: lib/statsd.js [hunk 1]
```diff
@@ -402,7 +348,11 @@ Client.prototype.sendMessage = function (message, callback) {
...
   }
 
-  if (this.protocol === TCP && message.lastIndexOf('\n') !== message.length - 1) {
+  if (!this.socket) {
+    return callback(new Error('No transport set. Check previous errors for details.'));
+  }
+
+  if (this.protocol === PROTOCOL.TCP && message.lastIndexOf('\n') !== message.length - 1) {
     message += '\n';
   }
```
Comment (msiebuhr) [inline lib/statsd.js]: I think this would be a better fit inside lib/transport.js

File: lib/statsd.js [hunk 2]
```diff
@@ -527,19 +470,7 @@ Client.prototype._close = function (callback) {
...
-      // this close is synchronous, and the socket will not emit a
-      // close event, so we can callback right away
-      if (callback) {
-        callback();
-      }
-    }
-    else {
-      this.socket.close();
-    }
+    this.socket.close();
```
Comment (msiebuhr) [inline lib/statsd.js]: How about also setting `this.socket = undefined`, so there's no doubt things are broken?

(And then it's trivial to do `if (!this.socket) { this.socket = createSocket(...) };` to attempt self-healing later on :)

File: .prettierrc [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+  "singleQuote": true
+}
```
Comment (msiebuhr) [inline .prettierrc]: While I can't speak for the project, random dotfiles is a smell to me. AFAIK, this project doesn't use https://prettier.io/.

Comment (runk) [reply .prettierrc]: Yeah, my bad - gonna kill it.

File: lib/statsd.js (hunk 1 above)
Comment (runk) [reply lib/statsd.js]: Good point!

File: lib/statsd.js (hunk 2 above)
Comment (runk) [reply lib/statsd.js]: Well, I'm trying not to change the logic in this PR and leaving behaviour as is. I better leave it intact.

Another place that bugs me a lot is https://github.com/brightcove/hot-shots/pull/129/files#diff-6b52e4167496e1de6769211f454200f7R80 - basically swallowing errors sometimes. Wish we'd just threw it, at least at setup stage.

File: lib/statsd.js [hunk 4]
```diff
@@ -402,8 +348,8 @@ Client.prototype.sendMessage = function (message, callback) {
     return;
   }
 
-  if (this.protocol === TCP && message.lastIndexOf('\n') !== message.length - 1) {
```
Comment (runk) [inline lib/statsd.js]: This has been moved to the tcp transport adapter

File: lib/statsd.js [hunk 5]
```diff
@@ -160,7 +106,7 @@ const Client = function (host, port, prefix, suffix, globalize, cacheDns, mock,
   // will gracefully (attempt) to re-open the socket with a small delay
   // options.udsGracefulRestartRateLimit is the minimum time (ms) between creating sockets
   // does not support options.isChild (how to re-create a socket you didn't create?)
-  if (!options.isChild && options.protocol === 'uds' && options.udsGracefulErrorHandling) {
+  if (!options.isChild && options.protocol === PROTOCOL.UDS && options.udsGracefulErrorHandling) {
```
Comment (runk) [inline lib/statsd.js]: Would be nice to abstract this piece in the uds transport, but perhaps a story for another PR. This is very special case for uds only.

**Comment by bdeitte** [comment]
Thanks @runk!  @msiebuhr I can tell you that I very much welcome the idea of having more reviews of things and having other people who help maintain.  So always welcome to do this, thanks.  :)

File: lib/constants.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+exports.PROTOCOL = {
+  TCP: 'tcp',
+  UDS: 'uds',
+  UDP: 'udp',
+  STREAM: 'stream'
```
Comment (bdeitte) [inline lib/constants.js]: Please add stream to the README

File: lib/statsd.js [hunk 7]
```diff
@@ -402,8 +348,8 @@ Client.prototype.sendMessage = function (message, callback) {
     return;
   }
 
-  if (this.protocol === TCP && message.lastIndexOf('\n') !== message.length - 1) {
-    message += '\n';
+  if (!this.socket) {
+    return callback(new Error('No transport set. Check previous errors for details.'));
```
Comment (bdeitte) [inline lib/statsd.js]:
```suggestion
    return callback(new Error('No protocol set. Check previous errors for details.'));
```

Since protocol is still the name exposed to users

File: lib/statsd.js [hunk 8]
```diff
@@ -418,23 +364,16 @@ Client.prototype.sendMessage = function (message, callback) {
       if (this.errorHandler) {
         this.errorHandler(errFormatted);
       } else {
-        console.log(String(errFormatted));
+        console.error(String(errFormatted));
```
Comment (bdeitte) [inline lib/statsd.js]: Thanks for the extra fix here

File: lib/constants.js (hunk 6 above)
Comment (bdeitte) [reply lib/constants.js]: I understand now reading farther down that you haven't added this in yet as an actual option, that this would be a follow on PR or still coming

File: test/init.js [hunk 9]
```diff
@@ -231,22 +228,4 @@ describe('#init', () => {
     statsd = createHotShotsClient(['host', 1234, 'prefix', 'suffix', false, false, true], clientType);
     assert.ok(statsd.mock);
   });
-
-  it('should create a socket variable that is an instance of dgram.Socket', () => {
-    statsd = createHotShotsClient({}, clientType);
-    assert.ok(statsd.socket instanceof dgram.Socket);
-    skipClose = true;
-  });
```
Comment (bdeitte) [inline test/init.js]: Is there a reason this was removed?  In the past people did access this directly, for better or worse, to catch errors, so this would become a backwards-compatibility issue to lose.

File: lib/constants.js (hunk 6 above)
Comment (runk) [reply lib/constants.js]: yes, that's the plan

**Comment by bdeitte** [comment]
A couple comments.  I really like the direction of abstracting this out.

File: test/init.js (hunk 9 above)
Comment (runk) [reply test/init.js]: Well, I was unsure what to do since `socket` is not an instance of the net/dgram anymore.

**Comment by runk** [comment]
@bdeitte @msiebuhr can you please check it out again?

**Comment by bdeitte** [comment]
Thanks, looks good to me @runk.  I'll wait to see if @msiebuhr has any more comments and then can merge in.

File: test/helpers/helpers.js [hunk 10]
```diff
@@ -5,7 +5,7 @@ const fs = require('fs');
 const os = require('os');
 const process = require('process');
 const StatsD = require('../../lib/statsd.js');
-
+const { PROTOCOL: { TCP, UDP, UDS } } = require('../../lib/constants');
```
Comment (msiebuhr) [inline test/helpers/helpers.js]: I'd personally prefer `const { TCP, UDP, UDS } = require('../../lib/constants').PROTOCOL;` - I find multi-layered unwrapping needlessly hard to follow. KISS and all...

**Comment by runk** [comment]
@bdeitte should be ready to go
